### PR TITLE
add commissions and commission rates

### PR DESCRIPTION
**Pull request by SirSplasch**
add commission agent to contact and order
add responsible user to order
add order_user

### CODE REVIEW

File: src/Livewire/DataTables/CommissionList.php [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+    public array $columnLabels = [
+        'user.name' => 'Commission Agent',
+        'order.order_number' => 'Order Number',
+        'order_position.name' => 'Order Position',
+    ];
+
+    public function mount(): void
+    {
+        $attributes = ModelInfo::forModel($this->model)->attributes;
```
Comment (patrickweh) [inline src/Livewire/DataTables/CommissionList.php:31]: move to one line

File: src/Livewire/DataTables/CommissionList.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        return $builder->with([
+            'user:id,firstname,lastname',
+            'order:id,order_number',
+            'orderPosition:id,name',
+        ]);
+    }
+
+    public function itemToArray($item): array
+    {
+        $item->commission_rate = $item->commission_rate['commission_rate'];
```
Comment (patrickweh) [inline src/Livewire/DataTables/CommissionList.php:58]: better set a label for this and remove the itemToArray

File: src/Livewire/DataTables/CommissionList.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            ]
+        );
+    }
+
+    public function getBuilder(Builder $builder): Builder
+    {
+        return $builder->with([
+            'user:id,firstname,lastname',
+            'order:id,order_number',
+            'orderPosition:id,name',
```
Comment (patrickweh) [inline src/Livewire/DataTables/CommissionList.php:52]: you dont need this

File: src/Livewire/Features/CommissionRates.php [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+    protected string $model = CommissionRate::class;
+
+    public array $enabledCols = [
+        'category.name',
+        'product.name',
+        'commission_rate',
+    ];
+
+    public array $columnLabels = [
```
Comment (patrickweh) [inline src/Livewire/Features/CommissionRates.php:32]: not necessary with new data table

File: src/Livewire/Features/CommissionRates.php [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+                'commission_rate' => 'percentage',
+            ]
+        );
+    }
+
+    public function getBuilder(Builder $builder): Builder
+    {
+        return $builder->with([
+            'user:id,firstname,lastname,email',
+            'category:id,name',
```
Comment (patrickweh) [inline src/Livewire/Features/CommissionRates.php:98]: this is also not needed

File: src/Models/Commission.php [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    public function user(): BelongsTo
+    {
+        return $this->belongsTo(User::class);
+    }
+
+    public function getLabel(): ?string
+    {
+        return null;
```
Comment (patrickweh) [inline src/Models/Commission.php:40]: could be category or product + rate_percentage?

**Review by patrickweh — CHANGES_REQUESTED** [review]

File: src/Listeners/Order/OrderInvoiceAddedSubscriber.php [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        foreach ($orderPositions as $orderPosition) {
+            $commissionRateId = null;
+            if (! $orderPosition->product_id
+                && $defaultCommissionRateByContact
+            ) {
+                $commissionRateId = $defaultCommissionRateByContact->id;
+            } elseif ($orderPosition->product_id) {
+                switch (true) {
+                    case $commissionRateId = $contactCommissionRates
+                        ->where('product_id', $orderPosition->product_id)
```
[the remark this answers is not in the document]
Comment (SirSplasch) [reply src/Listeners/Order/OrderInvoiceAddedSubscriber.php:62]: The database query already orders by commission_rate, so the first entry should always be the highest rate possible

**Review by patrickweh — APPROVED** [review]

**Pull request merged**